### PR TITLE
Adding v1beta1 Ordering Semantics

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This is required for Watch.

Where applicable, the semantics are copied from Node and the Android SDK. The Unit test is copied 1:1 from the Node SDK.

### CODE REVIEW

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BasePath.java [hunk 1]
```diff
@@ -93,6 +94,24 @@ boolean isPrefixOf(B path) {
     return true;
   }
 
+  /**
+   * Compare the current path against another Path object.
```
Comment (pongad) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BasePath.java]: Maybe also note that this is lexographic ordering?

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/Order.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        case REFERENCE_VALUE:
+          return REF;
+        case GEO_POINT_VALUE:
+          return GEO_POINT;
+        case ARRAY_VALUE:
+          return ARRAY;
+        case MAP_VALUE:
+          return OBJECT;
+        default:
+          throw new IllegalStateException("Could not detect value type for " + value);
```
Comment (pongad) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/Order.java]: IllegalArugmentException?

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/Order.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+   * @return -1 is left < right, 0 if left == right, otherwise 1
+   */
+  public int compare(@Nonnull Value left, @Nonnull Value right) {
+    // First compare the types.
+    TypeOrder leftType = TypeOrder.fromValue(left);
+    TypeOrder rightType = TypeOrder.fromValue(right);
+    int typeComparison = leftType.compareTo(rightType);
+    if (typeComparison != 0) {
+      return typeComparison < 0 ? -1 : 1;
+    }
```
Comment (pongad) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/Order.java]: There are lots of these "double compares". Do we care that deeply that the returned values are only -1,0,or 1?

Comment (schmidt-sebastian) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/Order.java]: I rely on this in the test case. But it makes much more sense to convert the end result in the test case itself, so I moved that conversion there.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/Order.java (hunk 2 above)
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/Order.java]: Done

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BasePath.java (hunk 1 above)
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/BasePath.java]: Done

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/Order.java (hunk 3 above)
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/Order.java]: I moved this conversion to the test class.

**Comment by schmidt-sebastian** [comment]
Review addressed. Thanks.

**Comment by pongad** [comment]
LGTM; @garrettjonesgoogle should +2

**Pull request merged**